### PR TITLE
serialize font: to empty on non-default subprops

### LINKED ISSUE #15036: `font` shorthand should be serialized to empty when some of its subproperties have non-default value
There are many subproperties of `font` shorthand which cannot present in `font` shorthand (see #15033). When any of those properties have non-default value, `font` shorthand should be serialized to empty string.

### PR DESCRIPTION
**Pull request by zploskey**
Fixes font shorthand serialization so that it serializes to "" when non-default subproperties are defined. These subproperties are those defined in #15033.

Adds tests:
- font_should_serialize_to_empty_if_there_are_nondefault_subproperties
- font_should_serialize_all_available_properties

The second test was previously commented out and underwent some cleanup to make it run.

---

- [x] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #15036

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15604)

### CODE REVIEW

File: components/style/properties/shorthand/font.mako.rs
Comment (zploskey) [inline components/style/properties/shorthand/font.mako.rs]: Would something like this be preferable?
```
                if let Some(val) = extract_value(self.${sub_property}) {
                    try!(val.to_css(dest));
                    try!(write!(dest, " "));
                }
```

Comment (nox) [inline components/style/properties/shorthand/font.mako.rs]: Yes.

**Comment by zploskey** [comment]
These latest changes fix some problems with my first attempt, and considerably simplify the changes needed to to_css_declared(). I believe the handling of non-default subproperties now meets the requirements. Unless the mako variable product is somehow set to "gecko" or "none", it is not really possible for the `should_serialize_to_empty_if_there_are_nondefault_subproperties()` test to work properly. It passes in its current form, but I believe that is only a coincidence since when I change the test to only serialize a property that is part of the font shorthand it also passes. So clearly the test doesn't work correctly right now.

What I can do is fix the test so that it will fail with the current test configuration, but we may want to either disable it or somehow mark it as a known failure. I see some documentation about annotating with "ignore" but nothing about known failures... maybe `should_panic` fills this purpose but doesn't seem to really describe my meaning. That being that the test will fail until the build system in test can set product to "gecko" or "none".

If there are no objections, I'll go ahead and fix the test by switching it to have at least one normal property of the font shorthand, call `shorthand_properties_to_string(properties)` and assert that the result == "". The test fails but I can annotate with `#[ignore]` for now. In the event that we eventually can build with product set to those values we would want to remove the ignore. Does this sound reasonable?

**Comment by zploskey** [comment]
Went ahead and adjusted the test as I described and ignored it. When enabled it fails for me with:

thread 'properties::serialization::shorthand_serialization::font::should_serialize_to_empty_if_there_are_nondefault_subproperties' panicked at 'assertion failed: `(left == right)` (left: `"font-style: italic; font-language-override: normal; font-kerning: none;"`, right: `""`)', /home/zach/src/servo/tests/unit/style/properties/serialization.rs:660

And now of course I just noticed that it is not producing a `font:` property at all, just the individual CSS properties. I guess that's expected. I was mistakenly thinking I could produce a font shorthand without adding all or most of its subproperties.

**Comment by zploskey** [comment]
Ok I think the test is now fixed to the extent that I think it can be. When enabled, it now fails with:

thread 'properties::serialization::shorthand_serialization::font::should_serialize_to_empty_if_there_are_nondefault_subproperties' panicked at 'assertion failed: `(left == right)` (left: `"font: italic normal bolder expanded 4px/3 serif; font-language-override: normal; font-kerning: none;"`, right: `"font-language-override: normal; font-kerning: none;"`)', /home/zach/src/servo/tests/unit/style/properties/serialization.rs:678

`font: ...` doesn't get remove for reasons mentioned previously. I've left the test ignored with a TODO about enabling it when have the capability to run those code paths in test.

I think that's all I can do for now. I'll wait for feedback.

**Comment by zploskey** [comment]
Changes have been rebased on master, intermediate commits squashed, and test changes split into a separate commit. Moved comment on one of the test functions above its annotations. Made some improvements to the commit messages.

**Comment by jdm** [comment]
@SimonSapin Review ping.

**Comment by SimonSapin** [comment]
Sorry I only realized now, but I think there is a misunderstanding. When @upsuper in #15036 writes "The `font` shorthand should be serialized to empty string", it means that the empty string should be returned *when requesting the value of the `font` property by itself* hrough the CSSOM method [`CSSStyleDeclaration.getPropertyValue`](https://drafts.csswg.org/cssom/#dom-cssstyledeclaration-getpropertyvalue). In our implementation this is the Rust method `PropertyDeclarationBlock::single_value_to_css`.

The tests in this PR are based on the Rust method `PropertyDeclarationBlock::to_css_string`, which implements the getter for the CSSOM attribute [`CSSStyleDeclaration.cssText`](https://drafts.csswg.org/cssom/#dom-cssstyledeclaration-csstext).

Although the two algorithms have some parts in common, they are different. They should be tested separately.

When asking for the value of a specific shorthand, there may be no syntax that would expand to the current values of all the corresponding longhands. In that case we return an empty string. When serializing an entire block, however, there is no such limitations since (in the general case) we’re serializing multiple declarations anyway. When a shorthand doesn’t have appropriate syntax, we can fall back to multiple longhand declarations instead. In no case should a (set of) declarations be entirely skipped from the serialization.

The links above lead to the parts of the CSSOM specification that specify these algorithms, including the fallback case:

Part of https://drafts.csswg.org/cssom/#serialize-a-css-declaration-block
> * Let *value* be the result of invoking serialize a CSS value of *current longhands*.
> * If *value* is the empty string, continue with the steps labeled *shorthand loop*.

**Comment by zploskey** [comment]
@SimonSapin Thanks for the clarification. My apologies for misunderstanding, this is my first PR on servo.

If I change the test to the following, attempting to use `single_value_to_css()` to serialize, it fails each time on the first assert because `x` is `Err(Error)`. It's unclear to me why that is the case.
```
        fn font_should_serialize_to_empty_if_there_are_nondefault_subproperties() {
            let declarations = vec![
                (PropertyDeclaration::FontStyle(DeclaredValue::Value(FontStyle::italic)),
                    Importance::Normal),
                (PropertyDeclaration::FontKerning(DeclaredValue::Value(FontKerning::none)),
                    Importance::Normal),
            ];

            let block = PropertyDeclarationBlock {
                declarations: declarations,
                important_count: 0
            };

            let mut s = String::new();

            let id = PropertyId::parse("font".into()).unwrap();
            let x = block.single_value_to_css(&id, &mut s);

            assert_eq!(x.is_ok(), true);
            assert_eq!(s, "");
        }
```
The approach is similar to `scroll_snap_type::should_serialize_to_empty_string_if_sub_types_not_equal()`. I actually tried this approach before but was never able to get it to work. I may be missing something basic about how `single_value_to_css()` is supposed to work. It fails even if I set only the first `PropertyDeclaration`.

**Comment by zploskey** [comment]
> Sorry I only realized now, but I think there is a misunderstanding. When @upsuper in #15036 writes "The font shorthand should be serialized to empty string", it means that the empty string should be returned when requesting the value of the font property by itself hrough the CSSOM method CSSStyleDeclaration.getPropertyValue. In our implementation this is the Rust method `PropertyDeclarationBlock::single_value_to_css`.

Do you mean `PropertyDeclarationBlock::property_value_to_css`? It has a that same link in its doc comment and it seems to work in this case while `single_value_to_css` does not because I need to pass more than a single PropertyDeclaration.

**Comment by zploskey** [comment]
Ignored test `font_should_serialize_to_empty_if_there_are_nondefault_subproperties` now fails with
>panicked at 'assertion failed: `(left == right)` (left: `"italic normal bolder expanded 4px/3 serif"`, right: `""`)', /home/zach/src/servo/tests/unit/style/properties/serialization.rs:687

because the products that would let this test pass are not enabled in the unit tests.

**Comment by zploskey** [comment]
PR was rebased on master. Changes in #15779 made the fn `to_css` rather than `to_css_declared`. Because `LonghandsToSerialize` now contains fields of type `DeclaredValue`, I have eliminated fn `extract_value` and its use. Because `to_css` returns a `fmt::Result`, I've adjusted it to return `Ok(())` when we encounter subproperties that are not part of the font shorthand.

@SimonSapin This is ready for review again.

File: components/style/properties/shorthand/font.mako.rs
Comment (nox) [inline components/style/properties/shorthand/font.mako.rs]:
```rust
if self.${sub_property}.is_some() { ... }
```

File: components/style/properties/shorthand/font.mako.rs
Comment (nox) [inline components/style/properties/shorthand/font.mako.rs]:
```rust
if self.font_language_override.is_some() { ... }
```

File: tests/unit/style/properties/serialization.rs
Comment (nox) [inline tests/unit/style/properties/serialization.rs]: Why change that?

**Review by nox — CHANGES_REQUESTED** [review]
A few nits.

File: tests/unit/style/properties/serialization.rs
Comment (zploskey) [inline tests/unit/style/properties/serialization.rs]: @nox It looks like you removed this comment, but in case it's not clear to others now that I've addressed the other nits: This did not compile using `atom!`, because the`string_cache` crate is not available, but this variation works. This test was commented out before, so I just uncommented and got it to compile by using `Atom::from` based on a suggestion from someone in IRC.

Comment (nox) [inline tests/unit/style/properties/serialization.rs]: I removed it because I thought you had reverted the change, but I'm ok with your explanation and we don't need `atom!` performance benefits here anyway.

**Comment by nox** [comment]
@SimonSapin This is ready to be reviewed.

**Comment by zploskey** [comment]
@SimonSapin I'd like to wrap this up soon before it has to be rebased again. Potential issues are that it checks whether nondefault properties are None rather than the result of the function returning the default. As far as I know they are None if they are not default. That is an easy fix if it is necessary. Otherwise, the main test for this PR can't be tested properly unless products "gecko" or "none" are enabled in the tests and it is therefore currently an ignored test. The other test just checks serialization of all the subproperties of the font shorthand and it currently passes. Thanks!

File: components/style/properties/shorthand/font.mako.rs
Comment (upsuper) [inline components/style/properties/shorthand/font.mako.rs]: I don't think this would build now. All subprops in `LonghandsToSerialize` are no longer `Option` after #15779, and basically they shouldn't be.

In this case, it should probably be something like
```rust
if self.font_${name} == font_${name}::get_initial_specified_value() {
    return Ok(());
}
```
but it is possible that not all of those longhands have `get_initial_specified_value()` method. In that case, you would need to add them. But most of them should be just single keyword properties, which would just work.

File: tests/unit/style/properties/serialization.rs
Comment (upsuper) [inline tests/unit/style/properties/serialization.rs]: Also this would not pass, because shorthand requires all its subprops to be present in the declaration block, including those it cannot represent.

You can take the test code of `transition` at the end of this file as an example of how to simplify this test code.

**Review by upsuper — CHANGES_REQUESTED** [review]

File: components/style/properties/shorthand/font.mako.rs
Comment (zploskey) [inline components/style/properties/shorthand/font.mako.rs]: It appears that all are single keyword properties or have `get_initial_specified_value()` so this change is all that's needed here. Thanks.

**Comment by zploskey** [comment]
I rewrote the all_available_properties test to use `parse_declaration_block()` as requested. Changes to `PropertyDeclarationBlock` required a rebase. I adjusted the nondefault_subproperties test to use `from_block()` instead. I think that addresses all the requested changes.

**Comment by SimonSapin** [comment]
@zploskey sorry for the delays here… If there’s a conflict again I can deal with the rebase, just make sure the [Allow edits from maintainers](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) checkbox is checked.

Looks good now, thanks!

@bors-servo r+

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[tests/unit/style/properties/serialization.rs, line 659 at r6](https://reviewable.io:443/reviews/servo/servo/15604#-Kf6mV8zmw13llImENLy:-Kf6mV8zmw13llImENLz:bwxgk2c) ([raw file](https://github.com/servo/servo/blob/bbc1cb227412ac45de4bc3634b2994f79fd78934/tests/unit/style/properties/serialization.rs#L659)):*
> ```Rust
>             ];
> 
>             let block = block_from(declarations);
> ```

Nit: please squash the "Use from_block" commit with the previous one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/15604#-:-Kf6klmybLtbxHskh6uc:b-ic60w6)*

**Comment by SimonSapin** [comment]
Several build errors like:
```rust
54273 |             if self.font_size_adjust == font_size_adjust::get_initial_specified_value() {
      |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::cmp::PartialEq<properties::longhands::font_size_adjust::SpecifiedValue>` is not implemented for `&'a properties::longhands::font_size_adjust::SpecifiedValue`
```

You can reproduce with `./mach build-geckolib`. The problem is that `self` contains `&Foo` references, while `get_initial_specified_value` returns `Foo`. The fix is to borrow the latter: `&font_size_adjust::get_initial_specified_value()`.

By the way the logic seems wrong, shouldn’t it be `!=` instead of `==`?

This also shows that `product == "none"` is never true, the code using that is dead. Instead try `% if data.testing:`. (And please add a test that fails without this bit.)

**Comment by zploskey** [comment]
@simonsapin wrote:
> You can reproduce with ./mach build-geckolib. The problem is that self contains &Foo references, while get_initial_specified_value returns Foo. The fix is to borrow the latter: &font_size_adjust::get_initial_specified_value().

Glad to find out how to build it with gecko enabled. Fixed.

> By the way the logic seems wrong, shouldn’t it be != instead of ==?

Yes, this apparently got mixed up in the review. Fixed.

> This also shows that product == "none" is never true, the code using that is dead. Instead try % if data.testing:. 

Awesome, knowing about data.testing really helps here actually. I've changed all the `"none"` checks to use `data.testing`. This seems to be the way to enable all these properties in the tests. The main test for this issue passes after enabling gecko-specific properties in testing, i.e. `% if product == "gecko" or data.testing`. This is what is currently done in the text shorthand, for example.

However, as a consequence of enabling `font_language_override` and/or the other product-specific properties in testing, `font_should_serialize_all_available_properties` now fails and just produces all the longhands instead of the font shorthand result we expect. I think it's related to the following,

@upsuper wrote:
> Also this would not pass, because shorthand requires all its subprops to be present in the declaration block, including those it cannot represent.

I think this is the issue I'm running into. Is there a workaround? If there's not an easy one, since this all_available_properties test is not directly the issue I was trying to fix, how do you feel about me just removing this test change from the PR? Everything else seems to work. It would be left commented as it was before.

@simonsapin wrote:
> (And please add a test that fails without this bit.)

Can you be a little more explicit about what you're thinking of for this proposed test? What would be the success and failure conditions?

**Comment by SimonSapin** [comment]
What I meant is that this line of Rust code ended up not being included due to a Mako/Python template bug, and yet the tests still passed. This means that either this line is useless, or the test do not exercise the cases where it’s useful. The latter means that the test coverage can be improved. To improved coverage, add a test that fails when the line is commented out, and passes when it’s included (with the template bug fixed). In this case, the line in question makes the serialization of `font` be the empty string when `font-language-override` has a value that is not the initial value. So this new test should include a non-initial value for `font-language-override`, but not anything else that could causes the the serialization of `font` to be empty, then assert that the serialization is indeed empty.

However, since I wrote that previous comment we’ve been discussing moving away from unit tests written in Rust for the `style` crate. These tests are tightly coupled with internal APIs and data structure that we’re refactoring all the time, so they need to be ported as well. Eventually I’d like to replace these unit tests with test written in JavaScript based on CSSOM. In addition to being less of a pain when refactoring internals, we can share such tests with Firefox and other browsers.

All of this to say that this PR looks good now, and there is no need to spend much time improving unit tests that we might remove/replace later.

@bors-servo r+

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/15604#-:-KfN7-tSk-9-xZu6XCrt:b7n8yxp)*

**Comment by SimonSapin** [comment]
> ./mach test-unit
>	thread 'properties::serialization::shorthand_serialization::font::font_should_serialize_all_available_properties' panicked at 'assertion failed: `(left == right)` (left: `"font-style: italic; font-variant: normal; font-weight: bolder; font-stretch: expanded; font-size: 4px; line-height: 3; font-family: serif;"`, right: `"font: italic normal bolder expanded 4px/3 serif;"`)', /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:654

**Comment by zploskey** [comment]
Yeah, I mentioned this problem in my last comment. For some reason adding these additional properties causes the base properties not to serialize to the font shorthand. If I do a find replace s/data.testing/not data.testing/ then that test passes. I'm having a hard time tracking down where the problem might be, but it's not in the function I've been working on, unless these properties need to somehow be marked as not required to serialize the shorthand.

**Comment by zploskey** [comment]
The test passes if we give it every single subproperty. Not sure if it's intended that the shorthand won't serialize otherwise, but this case seems to work.

**Comment by SimonSapin** [comment]
> The test passes if we give it every single subproperty.

That sounds right. A shorthand declaration is always equivalent to declarations for all of its longhands (with initial values for parts unspecified). So conversely, only if they’re all these is using a shorthand serialization equivalent.

@bors-servo r+

---

Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/15604#-:-Kf_cP5MFLJ2mogFpwtb:b-tvb42)*

**Comment by SimonSapin** [comment]
The two not-already-marked-intermittent failures match error messages and stack traces in #13509 and #14323. (@jdm, is this the right way to tag things, with the list of filenames removed from #14323?)

@bors-servo retry

<details>
<pre>
  ▶ TIMEOUT [expected PASS] /html/dom/elements/global-attributes/dir_auto-textarea-N-EN.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ Shutting down the Constellation after generating an output file or exit flag specified
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ failed to receive response to font request: IoError(Error { repr: Os { code: 104, message: "Connection reset by peer" } }) (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at /checkout/src/libcore/result.rs:860)
  │ stack backtrace:
  │    0:     0x7fc9a0662c2d - backtrace::backtrace::trace::hce6a3819d897a887
  │    1:     0x7fc9a0663112 - backtrace::capture::Backtrace::new::hb5a725a088a2a2fc
  │    2:     0x7fc99f586523 - servo::main::{{closure}}::h9d96b5c32a8afffa
  │    3:     0x7fc9a147c6bd - std::panicking::rust_panic_with_hook
  │                         at /checkout/src/libstd/panicking.rs:550
  │    4:     0x7fc9a147c554 - std::panicking::begin_panic<collections::string::String>
  │                         at /checkout/src/libstd/panicking.rs:511
  │    5:     0x7fc9a147c489 - std::panicking::begin_panic_fmt
  │                         at /checkout/src/libstd/panicking.rs:495
  │    6:     0x7fc9a147c417 - std::panicking::rust_begin_panic
  │                         at /checkout/src/libstd/panicking.rs:471
  │    7:     0x7fc9a14a947d - core::panicking::panic_fmt
  │                         at /checkout/src/libcore/panicking.rs:69
  │    8:     0x7fc9a038f51b - core::result::unwrap_failed::hd6c954846ed51283
  │    9:     0x7fc9a03a4a28 - gfx::font_cache_thread::FontCacheThread::last_resort_font_template::h1802e896535d574d
  │   10:     0x7fc9a03a5d87 - gfx::font_context::FontContext::layout_font_group_for_style::he3ec5275d9cc8fd2
  │   11:     0x7fc9a02ebf75 - layout::text::TextRunScanner::scan_for_runs::h9e7b4d5cef563b9e
  │   12:     0x7fc99f846c18 - <layout::construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode>>::build_flow_for_block_starting_with_fragments::heca9cf2293a81bc4
  │   13:     0x7fc99f842f7c - <layout::construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode>>::build_flow_for_block_like::h3f6021bd678bdf31
  │   14:     0x7fc99f8425e0 - <layout::construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode>>::build_flow_for_block::h99cf3ea7a8a001fe
  │   15:     0x7fc99f7f2e20 - <layout::construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode> as layout::traversal::PostorderNodeMutTraversal<ConcreteThreadSafeLayoutNode>>::process::h7163180639338351
  │   16:     0x7fc99f7e36be - <layout::traversal::RecalcStyleAndConstructFlows as style::traversal::DomTraversal<E>>::process_postorder::hb395296ca2e298d3
  │   17:     0x7fc99f81fbb7 - style::sequential::traverse_dom::doit::h473529d272cb7d1d
  │   18:     0x7fc99f81fb86 - style::sequential::traverse_dom::doit::h473529d272cb7d1d
  │   19:     0x7fc99f81fb86 - style::sequential::traverse_dom::doit::h473529d272cb7d1d
  │   20:     0x7fc99f81fb86 - style::sequential::traverse_dom::doit::h473529d272cb7d1d
  │   21:     0x7fc99f85dbb8 - layout_thread::LayoutThread::handle_reflow::hbab16fcce01ae6f3
  │   22:     0x7fc99f856495 - layout_thread::LayoutThread::handle_request_helper::h6861378e293d9e98
  │   23:     0x7fc99f852218 - <layout_thread::LayoutThread as layout_traits::LayoutThreadFactory>::create::{{closure}}::he70c89f1fd32aa44
  │   24:     0x7fc99f804487 - std::panicking::try::do_call::h761cbb26c4556062
  │   25:     0x7fc9a14836da - panic_unwind::__rust_maybe_catch_panic
  │                         at /checkout/src/libpanic_unwind/lib.rs:98
  │   26:     0x7fc99f81a3d7 - <F as alloc::boxed::FnBox<A>>::call_box::h1a58b1b3d97e5f50
  │   27:     0x7fc9a147b2a4 - alloc::boxed::{{impl}}::call_once<(),()>
  │                         at /checkout/src/liballoc/boxed.rs:650
  │                          - std::sys_common::thread::start_thread
  │                         at /checkout/src/libstd/sys_common/thread.rs:21
  │                          - std::sys::imp::thread::{{impl}}::new::thread_start
  │                         at /checkout/src/libstd/sys/unix/thread.rs:84
  │   28:     0x7fc99d450183 - start_thread
  │   29:     0x7fc99cf6737c - clone
  │   30:                0x0 - <unknown>
  │ ERROR:servo: failed to receive response to font request: IoError(Error { repr: Os { code: 104, message: "Connection reset by peer" } })
  │ called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }" (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at /checkout/src/libcore/result.rs:860)
  │ stack backtrace:
  │    0:     0x7fc9a0662c2d - backtrace::backtrace::trace::hce6a3819d897a887
  │    1:     0x7fc9a0663112 - backtrace::capture::Backtrace::new::hb5a725a088a2a2fc
  │    2:     0x7fc99f586523 - servo::main::{{closure}}::h9d96b5c32a8afffa
  │    3:     0x7fc9a147c6bd - std::panicking::rust_panic_with_hook
  │                         at /checkout/src/libstd/panicking.rs:550
  │    4:     0x7fc9a147c554 - std::panicking::begin_panic<collections::string::String>
  │                         at /checkout/src/libstd/panicking.rs:511
  │    5:     0x7fc9a147c489 - std::panicking::begin_panic_fmt
  │                         at /checkout/src/libstd/panicking.rs:495
  │    6:     0x7fc9a147c417 - std::panicking::rust_begin_panic
  │                         at /checkout/src/libstd/panicking.rs:471
  │    7:     0x7fc9a14a947d - core::panicking::panic_fmt
  │                         at /checkout/src/libcore/panicking.rs:69
  │    8:     0x7fc9a0275a39 - core::result::unwrap_failed::h8f88b98f0a536369
  │    9:     0x7fc9a02d756d - <layout::query::LayoutRPCImpl as script_layout_interface::rpc::LayoutRPC>::pending_images::h91063c5b1c3f0432
  │   10:     0x7fc99fe37a59 - script::dom::window::Window::force_reflow::hc4df658a79834119
  │   11:     0x7fc99fe3accb - script::dom::window::Window::reflow::h394a9997bedf7be7
  │   12:     0x7fc99fcacddf - script::dom::document::Document::finish_load::h1d8d2d557ecf4cc8
  │   13:     0x7fc99fdd86b3 - script::dom::servoparser::ServoParser::do_parse_sync::h2676270b9bff96b9
  │   14:     0x7fc99fdd806c - script::dom::servoparser::ServoParser::parse_sync::h4a7419b789141155
  │   15:     0x7fc99fde096a - <script::dom::servoparser::ParserContext as net_traits::FetchResponseListener>::process_response_eof::h0e6d7b63846ee146
  │   16:     0x7fc99fe6447a - <script::network_listener::ListenerRunnable<A, Listener> as script::script_thread::Runnable>::handler::h9c76f08103f95c48
  │   17:     0x7fc99fe872b5 - script::script_thread::ScriptThread::handle_msg_from_script::h4fb921a02f36cead
  │   18:     0x7fc99fe7ea8c - script::script_thread::ScriptThread::handle_msgs::{{closure}}::h95a96e772221ddd7
  │   19:     0x7fc99fe79d5a - script::script_thread::ScriptThread::handle_msgs::hb7c261538c9b5201
  │   20:     0x7fc99fe746c7 - script::script_thread::ScriptThread::start::hb46c4910de904afb
  │   21:     0x7fc99f9e0914 - std::panicking::try::do_call::ha18c2fc1e2f674e1
  │   22:     0x7fc9a14836da - panic_unwind::__rust_maybe_catch_panic
  │                         at /checkout/src/libpanic_unwind/lib.rs:98
  │   23:     0x7fc99fae5597 - <F as alloc::boxed::FnBox<A>>::call_box::h667d881caee9bfc3
  │   24:     0x7fc9a147b2a4 - alloc::boxed::{{impl}}::call_once<(),()>
  │                         at /checkout/src/liballoc/boxed.rs:650
  │                          - std::sys_common::thread::start_thread
  │                         at /checkout/src/libstd/sys_common/thread.rs:21
  │                          - std::sys::imp::thread::{{impl}}::new::thread_start
  │                         at /checkout/src/libstd/sys/unix/thread.rs:84
  │   25:     0x7fc99d450183 - start_thread
  │   26:     0x7fc99cf6737c - clone
  │   27:                0x0 - <unknown>
  └ ERROR:servo: called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"
  ▶ CRASH [expected OK] /navigation-timing/nav2_test_open_data_uri.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ ERROR:script::dom::bindings::error: Error at data:text/html;charset=utf-8,%3C%21DOCTYPE%20html%3E%0D%0A%3Ctitle%3Edata%20URL%20source%20for%20navigation-timing%2Fnav2_test_open_data_uri.html%3C%2Ftitle%3E%0D%0A%3C%21--%20NB%3A%20this%20file%20isn%27t%20actually%20used%20any%20where%21%20--%3E%0D%0A%3Clink%20rel%3D%22author%22%20title%3D%22Google%22%20href%3D%22http%3A%2F%2Fwww.google.com%2F%22%20%2F%3E%0D%0A%3Cscript%3E%0D%0Avar%20observer%20%3D%20new%20PerformanceObserver%28%0D%0A%20%20%20%20function%20%28entryList%29%20%7B%0D%0A%20%20%20%20%20%20%20%20parent.postMessage%28%22observed%22%2C%20%22%2A%22%29%3B%0D%0A%20%20%20%20%20%20%20%20observer.disconnect%28%29%3B%0D%0A%20%20%20%20%7D%29%3B%0D%0Aobserver.observe%28%7BentryTypes%3A%20%5B%22navigation%22%5D%7D%29%3B%0D%0A%3C%2Fscript%3E%0D%0A:6:5 PerformanceObserver is not defined
  │ called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: UnexpectedEof, error: StringError("failed to fill whole buffer") }) } (thread RenderBackend, at /checkout/src/libcore/result.rs:860)
  │ stack backtrace:
  │    0:     0x7f93640e7c2d - backtrace::backtrace::trace::hce6a3819d897a887
  │    1:     0x7f93640e8112 - backtrace::capture::Backtrace::new::hb5a725a088a2a2fc
  │    2:     0x7f936300b523 - servo::main::{{closure}}::h9d96b5c32a8afffa
  │    3:     0x7f9364f016bd - std::panicking::rust_panic_with_hook
  │                         at /checkout/src/libstd/panicking.rs:550
  │    4:     0x7f9364f01554 - std::panicking::begin_panic<collections::string::String>
  │                         at /checkout/src/libstd/panicking.rs:511
  │    5:     0x7f9364f01489 - std::panicking::begin_panic_fmt
  │                         at /checkout/src/libstd/panicking.rs:495
  │    6:     0x7f9364f01417 - std::panicking::rust_begin_panic
  │                         at /checkout/src/libstd/panicking.rs:471
  │    7:     0x7f9364f2e47d - core::panicking::panic_fmt
  │                         at /checkout/src/libcore/panicking.rs:69
  │    8:     0x7f936414b10d - core::result::unwrap_failed::h482e2085da99dab3
  │    9:     0x7f936419c017 - webrender::render_backend::RenderBackend::run::h586004cebbd294f6
  │   10:     0x7f936413e957 - std::panicking::try::do_call::h17d08941604712ad
  │   11:     0x7f9364f086da - panic_unwind::__rust_maybe_catch_panic
  │                         at /checkout/src/libpanic_unwind/lib.rs:98
  │   12:     0x7f9364155837 - <F as alloc::boxed::FnBox<A>>::call_box::h4111409bfcdae236
  │   13:     0x7f9364f002a4 - alloc::boxed::{{impl}}::call_once<(),()>
  │                         at /checkout/src/liballoc/boxed.rs:650
  │                          - std::sys_common::thread::start_thread
  │                         at /checkout/src/libstd/sys_common/thread.rs:21
  │                          - std::sys::imp::thread::{{impl}}::new::thread_start
  │                         at /checkout/src/libstd/sys/unix/thread.rs:84
  │   14:     0x7f9360ed5183 - start_thread
  │   15:     0x7f93609ec37c - clone
  │   16:                0x0 - <unknown>
  │ ERROR:servo: called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: UnexpectedEof, error: StringError("failed to fill whole buffer") }) }
  └ Pipeline failed in hard-fail mode.  Crashing!
</pre>
</details>

**Comment by SimonSapin** [comment]
#14323 again ¯\_(ツ)_/¯

**Comment by jdm** [comment]
@bors-servo: retry
* #14323

**Pull request merged**
